controller/src/main/java/io/pravega/controller/store/stream/Int96CounterImpl.java

```diff
@@ -56,32 +46,47 @@ public class ZkInt96Counter {
     private final AtomicInt96 counter;
     @GuardedBy("lock")
     private CompletableFuture<Void> refreshFutureRef;
-    private ZKStoreHelper storeHelper;
 
-    public ZkInt96Counter(ZKStoreHelper storeHelper) {
+
+    public Int96CounterImpl(final BiFunction<CounterInfo, BiConsumer<Int96, Int96>, CompletableFuture<Void>> function) {
         this.lock = new Object();
         this.counter = new AtomicInt96();
         this.limit = new AtomicInt96();
         this.refreshFutureRef = null;
-        this.storeHelper = storeHelper;
+        this.function = function;
     }
 
-    CompletableFuture<Int96> getNextCounter() {
+    @Override
+    public CompletableFuture<Int96> getNextCounter(OperationContext context) {
         CompletableFuture<Int96> future;
         synchronized (lock) {
             Int96 next = counter.incrementAndGet();
             if (next.compareTo(limit.get()) > 0) {
                 // ignore the counter value and after refreshing call getNextCounter
-                future = refreshRangeIfNeeded().thenCompose(x -> getNextCounter());
+                future = refreshRangeIfNeeded(context).thenCompose(x -> getNextCounter(context));
```

Comment:
Does this need to be done under lock and does getNextCounter need to run under the same lock? If it does then this code is incorrect as it may be run on a different thread later. If it does not then move it out from under the lock.

Reply:
Done as suggested..